### DISCUSSION TITLE
[아파치 카프카] - 카프카 특징

### DISCUSSION THREAD

**Discussion by Irisation23** [cs-collections/cs-collections-discussion#29] [k. 대규모 서비스에 필요한 기술] [answered]
> 📝 구성

> [1. 높은 처리량](#높은-처리량)
> [2. 확장성](#-확장성)
> [3. 영속성](#영속성)
> [4. 고가용성](#고가용성)

---

# 높은 처리량

- 카프카는 `프로듀서`가 `브로커`로 데이터를 보낼 때와 `컨슈머`가 브로커로부터 데이터를 받을때 모두 데이터를 묶어서 처리한다.

> 🤔 Q. 왜 카프카는 데이터를 묶어서 처리할까?
> > 많은 양의 데이터를 송수신할 때 맺어지는 네트워크 비용은 무시할 수 없다. - 원인
> > 네트워크 통신 횟수를 최소한으로 줄일수 있는 방법은 데이터를 `묶음 단위`로 `배치(batch)`를 이용해 처리할 수 있기 때문에 실시간 로그데이터를 처리하는데 적합하다. - 결과

- 파티션 단위를 통해 동일 목적의 데이터를 여러 파티션에 분배하고 데이터를 `병렬 처리`할 수 있다.
- 파티션의 개수만큼 컨슈머 개수를 늘려서 동일 시간당 데이터 처리량을 늘릴 수 있다.

---

# 확장성

- 가정: 급작스러운 트래픽 증가(e.g 특정 이벤트가 원인)가 발생 한다면 카프카는 어떤 조치를 취할 수 있는가?
- 답변: 카프카는 가변적 환경에서 안정적 `확장`이 가능하도록 설계되어있다.

> 🤔 Q. 어떻게 카프카는 가변적인 환경에 대비 하는가?
> > 데이터가 적을 때는 `카프카 클러스터`의 브로커를 최소한의 개수로 운영한다.
> > 데이터가 많아진다면 클러스터의 브로커 개수를 자연스럽게 늘려 `스케일 아웃(scale-out)` 할 수 있다. 반대 경우 `스케일 인(scale-in)`할 수 있다.

- 카프카의 스케일 아웃, 스케일 인 과정은 클러스터의 무중단 운영을 지원한다. 이 때문에 안정적인 운영이 가능하다.

---

# 영속성

> 🚨 영속성이란? 데이터를 생성한 프로그램이 종료되더라도 사라지지 않은 데이터의 특성을 뜻한다.

- 카프카는 다른 메시징 플랫폼과 달리 데이터를 메모리에 저장하지 않고 `파일 시스템`에 저장한다.

> 🤔 Q. 파일 시스템은 일반적으로 느리지 않나? 
> > 카프카는 운영체제 레벨에서 파일 시스템을 최대한 활용하는 방법을 적용하였다.
> > 운영체제에서는 파일 I/O 성능 향상을 위해 [페이지 캐시(page cache)영역](https://ddongwon.tistory.com/56)을 메모리에 따로 생성하여 사용한다.

- 디스크 기반의 파일 시스템(페이지 캐시)을 활용한 덕분에 브로커 애플리케이션이 장애 발생으로 종료되더라도 프로세스를 재시작하여 안전하게 데이터를 다시 처리할 수 있다.

---

# 고가용성

- 카프카 클러스터는 3개 이상의 서버들로 운영된다.
- 이 때문에 일부 서버에 장애가 발생 하더라도 무중단으로 안전하고 지속적으로 데이터를 처리할 수 있다.
- 클러스터로 이루어진 카프카는 데이터의 `복제(replication)` 를 통해 고가용성의 특징을 가지게 되었다.
  - 온프레미스 환경의 서버 랙 또는 퍼블릭 클라우드의 리전 단위 장애에도 데이터를 안전하게 복제 할 수 있는 `브로커 옵션`들이 준비되어 있다. 

> 🤔 Q. 왜 카프카 클러스터는 3개 이상으로 구축 하는가?
> > 일단 카프카 클러스를 1대를 운영 하는 것은 단일 장애지점이 될 수 있기 때문에 지양 해야한다.
> > 2대로 운영할 경우는 단일 장애지점을 벗어날 수는 있지만 브로커간에 데이터 복제의 시간차로 인해 일부 데이터가 유실될 수 있다.
> > 3대 이상으로 운영 되어야 할 이유에 대한 정리는 향후 정리할 예정 // TODO

---

**Comment by ahngilwoong** [comment, marked as answer]
잘 읽었습니다. 

저도 공부 할 때 클러스터 안에 replication 되는 브로커의 개수가 2,3개로 권장된다고 봤는데 이렇게 될 경우
2~3개중 1개의 브로커는 리더의 역할로 실제 Pub(프로바이더)에게 요청을 받고 Sub(컨슈머)에게 정상적으로 주는 메시지 브로커의 역할을하고
나머지 브로커들은 메세지 브로커로서의 기능을 하지 않으며 팔로워가 되어 리더의 토픽들을 복제하며 읽어오며 언제나 리더가 되어 서비스할 준비가 되있도록 복제만 하는 걸로 공부 했습니다. 그런데 사실 이 replication되는 서버가 하나씩 늘어날수록 그 서버가 차지해야하는 용량은 리더 서버보다 2배 3배
높게 설정해야한다고 합니다 그만큼 리소스를 빨아먹는건데

제가 이게 검색에 잘 안나와서 같이 고민 해보고 싶은 부분은

1. 3대 이상으로 운영 되야할 이유가 복제 서버를 만들수록 빨아 먹히는 이 리소스 때문이 큰지 아니면 Zookeeper에서 꼭 홀수의 서버들을 권장하는데 이 때문에 3대로 권장 하는건지
2.  Zookeeper에서 홀수의 서버를 권장하는데 (다수결으로 서비스가 정상인지 아닌지 판단하기 때문?) 그러면 클러스터의 노드(브로커 or 서버)를 2대로 하면 절대 안되는것이 아닌지

를 같이 한번 찾아보고 각자 글 작성해도 좋을 것 같습니다.

**Reply by Irisation23** [reply]
좋습니다 좋은 주제라고 생각이 됩니다.
아직 정리중인 부분이 많아서 차후 정리되는 글에 중복 내용이 있을 수도 있으니 
차근차근 밟아 보다가 다시 돌아와 해당 내용에 대해 정리해 보도록 하겠습니다.
